### PR TITLE
Spatial Dropout layer

### PR DESCRIPTION
**Pull request by iamshnoo**
Spatial Dropout Layer

To compare PyTorch/numpy implementation with mlpack/armadillo implementation of the functionality, check this [Google Colab notebook](https://colab.research.google.com/drive/13TbYROndyUb7xvY0AIXJ5Px87yhW0ewh?usp=sharing).

types_of_dropout

### CODE REVIEW

File: src/mlpack/methods/ann/layer/spatial_dropout.hpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+ public:
+  //! Create the SpatialDropout object.
+  SpatialDropout();
+  /**
+   * Create the SpatialDropout object using the specified parameters.
+   *
+   * @param size The number of channels of each input image.
+   * @param ratio The probability of each channel getting dropped.
+   */
+  SpatialDropout(size_t size, double ratio = 0.5);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/spatial_dropout.hpp:59]: Just to be consistent with the rest of the codebase I would use `const` for each parameter here.

File: src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  if (!reset)
+  {
+    batchSize = input.n_cols;
+    inputSize = input.n_rows / size;
+    reset = true;
+  }
+
+  if (deterministic)
+    output = input;
+
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp:63]: We can remove the empty line here.

File: src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    probabilities.fill(ratio);
+    ann::BernoulliDistribution<> bernoulli_dist(probabilities, false);
+    maskRow = bernoulli_dist.Sample();
+    mask = arma::repmat(maskRow, inputSize, 1);
+
+    for (size_t n = 0; n < batchSize; n++)
+    {
+      arma::mat& inputImage = inputTemp.slice(n);
+      arma::mat& outputImage = outputTemp.slice(n);
+      outputImage = inputImage % mask * scale;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp:82]: Maybe I missed something, but if this is the only place where we use `inputImage` and `outputImage` I would directly use `inputTemp.slice(n)` and `outputTemp.slice(n)`.

File: src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      batchSize, false, false);
+  arma::cube gTemp(const_cast<arma::mat&>(g).memptr(), inputSize, size,
+      batchSize, false, false);
+
+  for (size_t n = 0; n < batchSize; n++)
+  {
+    arma::mat& gyImage = gyTemp.slice(n);
+    arma::mat& gImage = gTemp.slice(n);
+
+    gImage = gyImage % mask * scale;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp:103]: See comment above.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 5]
```diff
@@ -4296,3 +4296,65 @@ TEST_CASE("TransposedConvolutionWeightInitializationTest", "[ANNLayerTest]")
...
+{
+  arma::mat input, output, gy, g, outputExpected, gExpected;
+
+  // Set the seed so that the second channel is dropped every time.
+  arma::arma_rng::set_seed(0);
+  SpatialDropout<> module(3, 0.2);
+
+  // Input is a batch of 2 images, each of size (2,2) and having 4 channels.
+  input << 0.4963 << 0.0885 << 0.7682 << 0.1320 << 0.3074 << 0.4901 << 0.6341
+      << 0.8964 << 0.4556 << 0.3489 << 0.6323 << 0.4017 << arma::endr;
```
Comment (zoq) [inline src/mlpack/tests/ann_layer_test.cpp:4313]: I think the values are from the pytorch implementation?

Comment (iamshnoo) [reply src/mlpack/tests/ann_layer_test.cpp:4313]: The expected output and expected g matrices are from the Pytorch implementation.  I have added a comment for that a few lines below this.
The input matrix and the gy matrix are filled with random values.

**Comment by iamshnoo** [comment]
In the test case, I am using ```arma::arma_rng::set_seed(0)``` so that the second channel can specifically be dropped out (as in the image above).  That allows me to directly compare the output with an expected matrix. This works when I test it on Mac and Linux. However, I noticed that the Windows build failed for the last commit when checking this. I think ```arma::arma_rng::set_seed(0)``` is not working the way I expect it to on the Windows machine. Any clues as to how I can go about resolving this? Thanks!

File: src/mlpack/tests/ann_layer_test.cpp (hunk 5 above)
Comment (zoq) [reply src/mlpack/tests/ann_layer_test.cpp:4313]: Thanks for the clarification.

**Comment by zoq** [comment]
> In the test case, I am using `arma::arma_rng::set_seed(0)` so that the second channel can specifically be dropped out (as in the image above). That allows me to directly compare the output with an expected matrix. This works when I test it on Mac and Linux. However, I noticed that the Windows build failed for the last commit when checking this. I think `arma::arma_rng::set_seed(0)` is not working the way I expect it to on the Windows machine. Any clues as to how I can go about resolving this? Thanks!

It's not guaranteed that the random number is the same across all systems, the only workaround I can see is to check all possible solutions, if that makes sense?

File: src/mlpack/methods/ann/layer/spatial_dropout.hpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  //! A boolean used to do some internal calculations once initially.
+  bool reset;
+
+  //! The number of images in the batch.
+  size_t batchSize;
+
+  //! The number of pixels in each feature map.
+  size_t inputSize;
+
+  //! If true dropout and scaling is disabled, see notes above.
```
Comment (sreenikSS) [inline src/mlpack/methods/ann/layer/spatial_dropout.hpp:147]: Which notes? I think I am missing something.

**Comment by sreenikSS** [comment]
This looks alright. I'll merge it if you don't have anything else to add.

**Review by zoq — COMMENTED** [review]
Can you update the https://github.com/mlpack/mlpack/blob/master/HISTORY.md as well?

**Review by sreenikSS — APPROVED** [review]

**Pull request merged**